service/raft/raft_group_registry.hh

```diff
@@ -49,6 +51,50 @@ struct raft_server_for_group {
     raft_rpc& rpc;
     raft_sys_table_storage& persistence;
     std::optional<seastar::future<>> aborted;
+    std::optional<lowres_clock::duration> default_op_timeout;
+};
+
+struct raft_timeout {
+    seastar::compat::source_location loc = seastar::compat::source_location::current();
+    std::optional<lowres_clock::duration> value;
+};
+
+class raft_operation_timeout_error : public std::runtime_error {
+    using std::runtime_error::runtime_error;
+};
+
+// A wrapper for raft::server that adds timeout support to the main raft methods.
+// If an operation doesn't finish in a specified amount of time, raft_operation_timeout_error
+// exception is thrown.
+// An instance can be obtained through raft_group_registry methods get_server_with_timeouts
+// and group0_with_timeouts.
+// Passing std::nullopt as a raft_timeout parameter to methods of this class is equivalent
+// to calling the original raft methods without timeout.
+// Passing raft_timeout{} means 'use default timeout for this group', which is taken
+// from raft_server_for_group::default_op_timeout. For group0 default_op_timeout
+// is set to 1 minute and can be overridden in tests through the group0-raft-op-timeout-in-ms injection.
+// A custom timeout can be passed via raft_timeout::value, it will take precedence over
+// the default timeout default_op_timeout.
+// If no default_op_timeout is configured for a group and raft_timeout{} is passed without
+// the value parameter set, on_internal_error will be triggered.
+//
+// Use methods of these class if the code is handling a client request
+// and the client expects a potential error. Don't use timeouts for background
+// fibers, such as topology coordinator, since they wouldn't add much value.
+// The only thing the background fiber can do with a timeout is to retry, and
```

Comment:
@gusev-p @gleb-cloudius I'm actually not sure about that.

For example suppose we're in the middle of bootstrap and coordinator tries to transition from join_group0 to write_both_read_old state.

We lost quorum.

Should the coordinator get stuck there forever? Wouldn't it be better if we had timeout for this group 0 command too? It makes sense to fail the operation instead of sleeping forever.

On the other hand, there is no way to rollback from the operation if we lost quorum. So when coordinator wants to transition back to normal state, then indeed a timeout no longer makes sense.

So I wonder if we should timeout on the forward path of topology sagas, up to the point-of-no-return. After point-of-no-return, or on the abort path, we would not timeout.

WDYT?

Reply:
> Should the coordinator get stuck there forever? Wouldn't it be better if we had timeout for this group 0 command too? It makes sense to fail the operation instead of sleeping forever.

IIUC, it wouldn't get stuck:
```
void fsm::tick_leader() {
    if (election_elapsed() >= ELECTION_TIMEOUT) {
        // 6.2 Routing requests to the leader
        // A leader in Raft steps down if an election timeout
        // elapses without a successful round of heartbeats to a majority
        // of its cluster; this allows clients to retry their requests
        // with another server.
        return become_follower(server_id{});
    }
```
the leadership will be lost and and topology coordinator fiber will be aborted

Reply:
Right. Do we print any useful message in the log if that happens?
And what happens to a bootstrapping node which gets stuck waiting for response if that happens?

Reply:
We have a fiber that prints about gaining/loosing leadership in Scylla now. May be it worth adding a log message about coordinator fiber termination as well.